library/crypto/sha512.c

```diff
@@ -260,22 +260,28 @@ sha512_file (int fd, off_t offset, size_t count, void *resblock)
           if (sum == BLOCKSIZE)
             break;
 
-          if (n == 0) // eof
-              goto process_partial_block;
+          if (n == 0) { /* eof */
+              if (total + sum < count) {
+                  free (buffer);
+                  return -ESPIPE;
+              } else {
+                  goto process_partial_block;
+              }
+          }
         }
 
-      /* Process buffer with BLOCKSIZE bytes.  Note that
-                        BLOCKSIZE % 128 == 0
-       */
-      if (total + BLOCKSIZE < count) {
-          total += BLOCKSIZE;
+      if (total + sum < count) {
```

Comment:
It should be checked at higher level, that record is within file boundaries, isn't it?
It is still possible that file is truncated for some reason during reading, but I doubt this is the case.

So, what does this fix? Is it broken record which spans outside datafile? Or it is some runtime file truncation by external entities?